accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,706 @@
+The breaking change is: **`typeof(delegate*<...>)` and `Type.GetType()` will return a `Type` instance instead of `IntPtr`**
+
+The fallout from this is that for any general-purpose reflection code that wants to support function pointers will likely need to call `Type.IsFunctionPointer` if there is special treatment for function pointers. 
+
+# Potential future features
+Various features not planned for 8.0; somewhat discussed earlier.
+
+## Support `Type.FullName` and `Type.AssemblyQualifiedName`
+Including `Type.GetType()` and similar APIs to parse that string to construct a function pointer type dynamically.
+
+## `System.Reflection.FunctionPointer` for `System.Object`-based invoke
+See the existing [System.Reflection.Pointer](https://docs.microsoft.com/dotnet/api/system.reflection.pointer) for reference. This would allow passing strongly-typed, validated function pointers when using `System.Object`-based invoke. Naming of Box\Unbox vs. constructors etc. is TBD.
+
+ This is a value type, not a reference type like `Pointer`, since newer proposed reflection APIs will not force boxing to occur.
+
+ Like `Pointer` there is no property to obtain the type since it is used in a temporary manner to manually marshal.
+
+```diff
+namespace System.Reflection
+{
++   public struct FunctionPointer
++   {
++       public static FunctionPointer Box(void* ptr, Type type);
++       public static void* Unbox (FunctionPointer ptr);
++   }
+}
+```
+
+## Support Type.GetMethod()
+If a user tries to retrieve a Method through [`Type.GetMethod()`](https://learn.microsoft.com/dotnet/api/system.type.getmethod) that has function pointer arguments specified by the `Type[]` passed into that method, the current design throws `NotSupportedException` since we don't have a good way to specify the custom modifiers for each parameter in the API (see below -- requires "MethodSignature").
+
+The workaround is to have the caller manually loop through the `MethodInfo`s returned from `Type.GetMethods()` and perform the filtering there.
+
+```diff
+    public abstract class Type
+    {
+        // This is used with 'Type.GetMethod(..., Type[] types)' to find a method with parameters that
+        // are function pointers. For a similar reference, see Type.MakeGenericSignatureType().
++       public static Type MakeFunctionPointerType(Type[] functionPointerArguments, Type returnType, bool isUnmanaged);
+
+        // For unmanaged methods, we need to specify the modopts, so something like the proposed MethodSignature below
++       public static Type MakeFunctionPointerType(MethodSignature functionPointerArguments);
+        // which is preferred over something like this:
+        public static Type MakeFunctionPointerType(
+            Type[] argumentTypes,
+            Type[,] argumentOptionalModifiers,
+            Type[,] argumentRequiredModifiers,
+            Type[] returnType,
+            Type[] returnTypeOptionalModifiers,
+            Type[] returnTypeRequiredModifiers,
```

Comment:
```suggestion
            Type returnType,
            Type[]? returnTypeOptionalModifiers,
            Type[]? returnTypeRequiredModifiers,
            Type[] argumentTypes,
            Type[][]? argumentOptionalModifiers,
            Type[][]? argumentRequiredModifiers,
```
To match existing methods like 
https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.methodbuilder.setsignature

Reply:
Thanks. Updated.